src/malli/core.cljc

```diff
@@ -905,57 +932,65 @@
            (-set [this key value] (-set-entries this key value))))))))
 
-(defn -ref-schema
-  ([]
-   (-ref-schema nil))
-  ([{:keys [lazy type-properties] :as opts}]
-   ^{:type ::into-schema}
-   (reify IntoSchema
-     (-into-schema [_ properties [ref :as children] {::keys [allow-invalid-refs] :as options}]
-       (-check-children! :ref properties children {:min 1, :max 1})
-       (when-not (-reference? ref)
-         (-fail! ::invalid-ref {:ref ref}))
-       (let [-memoize (fn [f] (let [value (atom nil)] (fn [] (or @value) (reset! value (f)))))
-             -ref (or (and lazy (-memoize (fn [] (schema (mr/-schema (-registry options) ref) options))))
-                      (if-let [s (mr/-schema (-registry options) ref)] (-memoize (fn [] (schema s options))))
-                      (when-not allow-invalid-refs
-                        (-fail! ::invalid-ref {:type :ref, :ref ref})))
-             children (vec children)
-             form (-create-form :ref properties children)]
-         ^{:type ::schema}
-         (reify
-           Schema
-           (-type [_] :ref)
-           (-type-properties [_] type-properties)
-           (-validator [_]
-             (let [validator (-memoize (fn [] (-validator (-ref))))]
-               (fn [x] ((validator) x))))
-           (-explainer [_ path]
-             (let [explainer (-memoize (fn [] (-explainer (-ref) (conj path 0))))]
-               (fn [x in acc] ((explainer) x in acc))))
-           (-transformer [this transformer method options]
-             (let [this-transformer (-value-transformer transformer this method options)
-                   deref-transformer (-memoize (fn [] (-transformer (-ref) transformer method options)))]
-               (-intercepting this-transformer (fn [x] (if-some [t (deref-transformer)] (t x) x)))))
-           (-walk [this walker path options]
-             (let [accept (fn [] (-inner walker (-ref) (into path [0 0]) (-update options ::walked-refs #(conj (or % #{}) ref))))]
-               (if (-accept walker this path options)
-                 (if (or (not ((-boolean-fn (::walk-refs options false)) ref)) (contains? (::walked-refs options) ref))
-                   (-outer walker this path [ref] options)
-                   (-outer walker this path [(accept)] options)))))
-           (-properties [_] properties)
-           (-options [_] options)
-           (-children [_] children)
-           (-parent [_] (-ref-schema opts))
-           (-form [_] form)
-           LensSchema
-           (-get [_ key default] (if (= key 0) (-pointer ref (-ref) options) default))
-           (-keep [_])
-           (-set [this key value] (if (= key 0) (-set-children this [value])
-                                                (-fail! ::index-out-of-bounds {:schema this, :key key})))
-           RefSchema
-           (-ref [_] ref)
-           (-deref [_] (-ref))))))))
+(def -ref-schema
+  (let [into-ref-schema
+        (memoize
```

Comment:
Doesn't this introduce a memory leak? All schemas are persisted forever in the memoization cache? Also, as options are part of the memoization and can can mutable state (just as recursion counters with gen), the same schema can appear multiple times in the cache,

Reply:
This was just a quick hack to make `=` (and thus set membership) work for refs and I did suspect it wouldn't fly. I suppose some more complicated or nonextensible solution is in order, any ideas? (Too bad the memoization cache is so naive, OTOH weak maps have other issues and so on...)

Reply:
Not sure how to do that elegantly, but few guesses:
* schema form is immutable, could that be used as equality (like there is `malli.util/equals`)
* there is already the local `-ref`, which is memoized for caching purposes.

Reply:
and yes, ´clojure.core/memoize` is awful.

Reply:
Yep, memoize bit me too once in clj-kondo... huge memory leak ;)